### PR TITLE
Generate stub

### LINKED ISSUE #17: Add a `clash gen` command
Add a `clash gen` command similar to https://github.com/kyuridenamida/atcoder-tools that generates the template for solving a clash in a particular language.

This could deal with https://github.com/Andriamanitra/clash/issues/13 by adding the extra text to the file.

### PR DESCRIPTION
**Pull request by ellnix**
This is where you witness my descent into madness.
Also:
- Fixes #17

# Update 2023-12-09:
Necessary parts are done, left to do:
- [x] Use correct case in variable names in stub (using the 'variable_name_format' key in `stub_config.toml`, so that the templates do not have to worry about case)
  - [X] Apply it to `Read`s
  - [x] https://github.com/ellnix/clash/issues/9
  - [x] https://github.com/ellnix/clash/issues/10
  - [x] https://github.com/ellnix/clash/issues/11
  - [x] https://github.com/ellnix/clash/issues/12
  - [x] https://github.com/ellnix/clash/issues/13
- [x] Refactor
  - [x] Replace `unwrap` and `expect` s with better messages
  - [X] Properly squeeze multiple newlines
  - [x] Make the formatter happy
- [X] Add stubs in more languages to find problems with implementation
- [X] Rethink how "type tokens" are added to the template
  - [X] Are there languages that need more than one arbitrary token per type?
- [x] Fix template formatting for consistency
  - [x] https://github.com/ellnix/clash/issues/8
  - [x] Smaller things
    - [x] https://github.com/ellnix/clash/issues/7
- [x] https://github.com/ellnix/clash/issues/5

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -97,6 +99,18 @@ fn cli() -> clap::Command {
                     \n (1) https://www.codingame.com/contribute/community"
                 )
         )
+        .subcommand(
+            Command::new("generate-stub")
```
Comment (daxida) [inline src/main.rs:152]: Not very fan of the long name for a command that I see myself using almost every clash. Atcoder uses gen (that @Andriamanitra didn't like I think). Maybe just generate, maybe just stub?

File: src/main.rs [hunk 2]
```diff
@@ -154,6 +168,13 @@ impl App {
         }
     }
 
+    fn programming_language_from_args(&self, args: &ArgMatches) -> Result<ProgrammingLanguage> {
+        match args.get_one::<String>("PROGRAMMING_LANGUAGE") {
+            Some(s) => Ok(ProgrammingLanguage::from(s.to_owned())),
+            None => Err(anyhow!("No clash handle given")),
+        }
+    }
```
Comment (daxida) [inline src/main.rs]: Unfortunate error message.

Comment (ellnix) [reply src/main.rs]: Woops, remember it's never the first copy pasta that get you, it's always the second one when the first doesn't work.

File: src/main.rs (hunk 1 above)
Comment (ellnix) [reply src/main.rs:152]: I don't really mind either way, feel free to discuss it with @Andriamanitra or in #17 and I'll change it accordingly.

Comment (Andriamanitra) [reply src/main.rs:152]: I'm not opposed to adding an [alias](https://docs.rs/clap/latest/clap/struct.Command.html#method.alias) called `gen` but for the purposes of --help and documentation I think the longer name is way better.

Comment (ellnix) [reply src/main.rs:152]: Not sure how to create an alias, I'll leave it for the end.

File: src/main.rs (hunk 2 above)
Comment (ellnix) [reply src/main.rs]: Should be fixed now.

File: src/main.rs (hunk 1 above)
Comment (daxida) [reply src/main.rs:152]: It's just: `.alias("gen")` after the command

File: src/stub.rs [hunk 3]
```diff
@@ -33,9 +33,9 @@ pub enum ReadStub {
...
-enum Stub {
-  Read(Vec<ReadStub>),
-  Loop(Box<Stub>),
-  LoopLine(Box<Stub>),
+pub enum Stub {
+  Read(Vec<VariableStub>),
+  Loop { count: String, command: Box<Stub> },
+  LoopLine { object: String, variables: Vec<VariableStub> },
   Write(String),
 }
```
Comment (daxida) [inline src/stub.rs]: You will need to add a Statement to that enum if you want to close #13 

This is clash that uses it: https://www.codingame.com/contribute/view/653884ada8d44d632a5f42168453bf565049a

File: src/clash.rs [hunk 4]
```diff
@@ -85,6 +85,10 @@ impl Clash {
         self.last_version.data.constraints.as_deref()
     }
 
+    pub fn stub_generator(&self) -> Option<&str> {
```
Comment (daxida) [inline src/clash.rs:88]: This is confusing, the clash doesn't have a stub_generator, it has a stub that then goes through the cg stub_generator to produce the default code. 

From https://www.codingame.com/playgrounds/40701/help-center/stub-generator:

> The stub generator is the program which will write in every available language the default code you get when you open the IDE. To do this, the generator needs input written using the syntax below.

So I think here it should really just be `data.stub.as_deref()` etc.

File: src/clash.rs [hunk 5]
```diff
@@ -85,6 +85,10 @@ impl Clash {
...
+        self.last_version.data.stub_generator.as_deref()
+    }
+
```
Comment (daxida) [inline src/clash.rs:91]: This is confusing, the clash doesn't have a stub_generator, it has a stub that then goes through the cg stub_generator to produce the default code.

From https://www.codingame.com/playgrounds/40701/help-center/stub-generator:

The stub generator is the program which will write in every available language the default code you get when you open the IDE. To do this, the generator needs input written using the syntax below.

So I think here it should really just be data.stub.as_deref() etc.

**Comment by ellnix** [comment]
Forgive us for our sins @Andriamanitra

File: src/main.rs [hunk 6]
```diff
@@ -147,6 +148,19 @@ fn cli() -> clap::Command {
                         .value_parser(value_parser!(PublicHandle))
                 )
         )
+        .subcommand(
+            Command::new("generate-stub")
+                .alias("gen")
+                .about("Generate input handling code for a given language")
+                .arg(arg!(<PROGRAMMING_LANGUAGE> ... "programming language of the solution stub"))
```
Comment (Andriamanitra) [inline src/main.rs]: The `...` probably shouldn't be there as it allows you to specify multiple programming languages (but the logic only handles the first one anyway)

File: src/stub/parser.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
+#![allow(clippy::while_let_on_iterator)]
+
+use regex::Regex;
+
+pub mod types;
+pub use types::{Cmd, InputComment, JoinTerm, JoinTermType, LengthType, Stub, VariableCommand};
+
+pub fn parse_generator_stub(generator: String) -> Stub {
+    let generator = generator.replace('\n', " \n ").replace("\n  \n", "\n \n");
```
Comment (Andriamanitra) [inline src/stub/parser.rs]: Why is this necessary?

File: src/stub/parser.rs [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        }
+
+        vars
+    }
+
+    fn parse_output_comment(&mut self, previous_commands: &mut [Cmd]) {
+        let output_comment = self.parse_statement();
+        for cmd in previous_commands {
+            Self::update_cmd_with_output_comment(cmd, &output_comment)
+        }
```
Comment (Andriamanitra) [inline src/stub/parser.rs:201]: Why does this need to loop over all previous commands?

File: src/stub/renderer/language.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+                match fs::read_to_string(language_config_filepath) {
+                    Ok(config_file_content) => toml::from_str::<Language>(&config_file_content).ok(),
+                    _ => None,
+                }
+            })
+            .find(|l| match &l.aliases {
+                Some(aliases) => aliases.contains(&name.to_string()),
+                None => false,
+            })
+    }
```
Comment (Andriamanitra) [inline src/stub/renderer/language.rs]: This should return a `Result` instead of `Option` so we can print better error messages (for example if the stub_config file is missing or the toml is invalid).

File: src/stub/renderer/types.rs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+#[derive(Debug, Clone, Serialize)]
+pub struct ReadData {
+    pub name: String,
+    pub var_type: VariableType,
+    pub max_length: Option<String>,
+    pub length_type: Option<LengthType>,
+}
+
+impl ReadData {
+    // VariableNameFormat is just the case (snake_case, pascal_case etc.)
```
Comment (Andriamanitra) [inline src/stub/renderer/types.rs]: Maybe it should be called `VariableNameCasing` or something so it doesn't need a comment to explain?

File: justfile [hunk 11]
```diff
@@ -49,3 +49,63 @@ check-nested-self:
...
+
+# Test the stub generator with a random clash in rust
+test-stub-rs:
+    cargo run -- next
+    cargo run -- generate-stub "rust"
+
+# Test the stub generator with a random clash in C
+test-stub-c:
+    cargo run -- next
+    cargo run -- generate-stub "c"
```
Comment (Andriamanitra) [inline justfile]: you can reduce repetition by using parameters:
```suggestion
test-stub LANG:
    cargo run -- next
    cargo run -- generate-stub {{LANG}}
```

File: src/stub/renderer/language.rs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+            .ok_or(anyhow!("Unsupported language: {}", value))
+    }
+}
+
+impl Language {
+    pub fn template_glob(&self) -> String {
+        format!("config/stub_templates/{}/*.{}.jinja", self.name, self.source_file_ext)
+    }
+
+    fn find_lang_by_name<'a>(name: &'a str, lang_folders: &'a [String]) -> Result<Option<Language>> {
```
Comment (Andriamanitra) [inline src/stub/renderer/language.rs]: I think it would be simpler to have this return `Result<Language>` ("not found" error instead of Ok(None))

**Review by Andriamanitra — CHANGES_REQUESTED** [review]
The stub Parser could really use some unit tests, both to prevent regressions and (maybe more importantly) as documentation of what it's supposed to be doing because its logic is a bit difficult to follow.

File: src/stub/parser.rs (hunk 7 above)
Comment (daxida) [reply src/stub/parser.rs]: In the future the second replace won't be needed. I haven't managed to get rid of the first one yet.

File: src/stub/renderer/types.rs (hunk 10 above)
Comment (daxida) [reply src/stub/renderer/types.rs]: I added this comment I think, because I didn't understand at first what was that struct doing.

File: justfile (hunk 11 above)
Comment (ellnix) [reply justfile]: @daxida I think you missed something you are guilty on :)

File: src/main.rs (hunk 6 above)
Comment (ellnix) [reply src/main.rs]: My bad, I think this is just copy pasta from the previous Commands

File: src/stub/renderer/language.rs (hunk 12 above)
Comment (ellnix) [reply src/stub/renderer/language.rs]: `Result` is an `Err` only if there was a language by the name provided but there was a problem reading the file. `Ok(None)` means that there simply wasn't a language by that exact match and that allows the code to continue searching for a language that is aliased to the parameter.

File: src/stub/renderer/types.rs (hunk 10 above)
Comment (ellnix) [reply src/stub/renderer/types.rs]: The comment is not necessary anymore all the same, VariableNameFormat is now only used in `Language` (and abstracted through `Language` too) where its purpose is clear from the methods defined on it.

File: src/stub/renderer/language.rs (hunk 9 above)
Comment (ellnix) [reply src/stub/renderer/language.rs]: Here the choice was that invalid toml would not be fatal, the app simply continues looking for a match for the alias, if they all fail then we return an error message.

If you are making a new language toml, you should first try gen with the language name, not with an alias, in which case you would run into:
```rust
            Ok(toml::from_str(&config_file_content)
                .context("There was an error loading the stub configuration")?)
```

It didn't make sense to me to raise an error when any language toml was invalid, since the command doesn't expressly state that it verifies that all configuration is valid.

If you disagree, I can change it (no hard feelings, obviously).

File: src/stub/parser.rs (hunk 7 above)
Comment (ellnix) [reply src/stub/parser.rs]: To answer the original question, the intent was to split on whitespace and then parse each token as they come, but I also needed to capture newlines for the parser (splitting consumes the match). Essentially I needed this transformation:
```rust
"ab c\n\nde" => ["ab", "c", "\n", "\n", "de"]
```
Which I get by these two replaces:
```rust
"ab c\n\nde" => 
"ab c \n  \n de" => 
"ab c \n \n de" => 
(split on space) ["ab", "c", "\n", "\n", "de"]
```

File: justfile (hunk 11 above)
Comment (daxida) [reply justfile]: Fixed in [7416270](https://github.com/Andriamanitra/clash/pull/31/commits/74162708894258608b8aa660b35dcc7d2705f710)

File: src/stub/renderer/types.rs (hunk 10 above)
Comment (daxida) [reply src/stub/renderer/types.rs]: `renderer::types.rs` doesn't exist anymore. Instead we use VariableCommand directly in jinja.

File: src/main.rs (hunk 6 above)
Comment (ellnix) [reply src/main.rs]: Solved

File: config/stub_templates/c/read_many.c.jinja [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+---- 🇷 EAD_MANY
+{%- if single_type -%} 
+  (SINGLE_TYPE)
+{% else -%}
+  (MULTIPLE_TYPE)
+{% endif -%}
+{% endif -%}
+
+{%- for var in vars %}
+{%- if var.input_comment -%}# {{ var.ident }}: {{ var.input_comment }}
```
Comment (Andriamanitra) [inline config/stub_templates/c/read_many.c.jinja]:
```suggestion
{%- if var.input_comment -%}// {{ var.ident }}: {{ var.input_comment }}
```

File: config/stub_templates/rust/loopline.rs.jinja [hunk 14]
```diff
@@ -0,0 +1,11 @@
+{%- if debug_mode -%}
+# LOOPLINE
+{% endif -%}
+
+gets.split.each{% if vars | length > 1 %}_slice({{ vars | length }}){% endif %} do |{{ vars | map(attribute="ident") | join(sep=", ") }}|
```
Comment (Andriamanitra) [inline config/stub_templates/rust/loopline.rs.jinja]: this isn't rust at all

File: config/stub_templates/rust/loopline.rs.jinja [hunk 15]
```diff
@@ -0,0 +1,43 @@
+{%- set vars_length = vars | length -%}
+
+{# SINGLE_TYPE #}
+{%- if vars_length == 1 -%} 
+{%- set var = vars[0] -%}
+
+{# Special CG case, doesn't make much sense otherwise #}
+{%- if var.var_type == "Word" -%}
```
Comment (Andriamanitra) [inline config/stub_templates/rust/loopline.rs.jinja]:
```suggestion
{%- if var.var_type == "Word" or var.var_type == "String" -%}
```
